eclair-core/src/main/scala/fr/acinq/eclair/wire/Onion.scala

```diff
@@ -17,49 +17,133 @@
 package fr.acinq.eclair.wire
 
 import fr.acinq.bitcoin.ByteVector32
-import fr.acinq.eclair.ShortChannelId
 import fr.acinq.eclair.crypto.Sphinx
-import scodec.bits.{BitVector, ByteVector}
+import fr.acinq.eclair.wire.CommonCodecs._
+import fr.acinq.eclair.wire.OnionTlv._
+import fr.acinq.eclair.wire.TlvCodecs._
+import fr.acinq.eclair.{ShortChannelId, UInt64}
+import scodec.bits.{BitVector, ByteVector, HexStringSyntax}
 import scodec.codecs._
 import scodec.{Codec, DecodeResult, Decoder}
 
 /**
   * Created by t-bast on 05/07/2019.
   */
 
+/**
+  * Tlv types used inside onion messages.
+  */
+sealed trait OnionTlv extends Tlv
+
 case class OnionRoutingPacket(version: Int,
                               publicKey: ByteVector,
                               payload: ByteVector,
                               hmac: ByteVector32)
 
-case class PerHopPayload(shortChannelId: ShortChannelId,
-                         amtToForward: Long,
-                         outgoingCltvValue: Long)
+case class OnionForwardInfo(shortChannelId: ShortChannelId,
+                            amtToForward: Long,
+                            outgoingCltvValue: Long)
+
+case class OnionPaymentInfo(amountMsat: Long, cltvExpiry: Long)
+
+case class OnionPerHopPayload(payload: Either[TlvStream[OnionTlv], OnionForwardInfo]) {
+
+  lazy val paymentInfo: Option[OnionPaymentInfo] = payload match {
+    case Right(OnionForwardInfo(_, amount, cltv)) => Some(OnionPaymentInfo(amount, cltv))
+    case Left(tlv) =>
+      tlv.records.foldLeft((None, None): (Option[Long], Option[Long])) {
+        case ((_, cltv), AmountToForward(amount)) => (Some(amount), cltv)
+        case ((amount, _), OutgoingCltv(cltv)) => (amount, Some(cltv))
+        case (res, _) => res
+      } match {
+        case (Some(amount), Some(cltv)) => Some(OnionPaymentInfo(amount, cltv))
+        case _ => None
+      }
+  }
+
+  lazy val forwardInfo: Option[OnionForwardInfo] = payload match {
+    case Right(onionForwardInfo) => Some(onionForwardInfo)
+    case Left(tlv) =>
+      tlv.records.foldLeft((None, None, None): (Option[ShortChannelId], Option[Long], Option[Long])) {
+        case ((_, amount, cltv), OutgoingChannelId(channelId)) => (Some(channelId), amount, cltv)
+        case ((channelId, _, cltv), AmountToForward(amount)) => (channelId, Some(amount), cltv)
+        case ((channelId, amount, _), OutgoingCltv(cltv)) => (channelId, amount, Some(cltv))
+        case (res, _) => res
+      } match {
+        case (Some(channelId), Some(amount), Some(cltv)) => Some(OnionForwardInfo(channelId, amount, cltv))
+        case _ => None
+      }
+  }
+
+}
+
+object OnionPerHopPayload {
+
+  // @formatter:off
+  implicit def legacyPayloadToPerHopPayload(legacy: OnionForwardInfo): OnionPerHopPayload = OnionPerHopPayload(Right(legacy))
```

Comment:
I'm not a big fan of this, do you have a suggestion on how to improve this?

Reply:
You could override `OnionPerHopPayload.apply()` and use `OnionPerHopPayload(...)` explicitely but it's not really "lean"

Reply:
Yeah it would force us to always use the `OnionPerHopPayload` constructor explicitly which is a bit cumbersome...could I somehow define an extension method `toOnionPerHopPayload` for `OnionForwardInfo` and `TlvStream[OnionTlv]`?